File: ts/src/contrast.ts
/**
 * WCAG Contrast & Relative Luminance Utilities
 *
 * Implements WCAG 2.1 contrast ratio calculation and automatic
 * accessible text color selection (White or Black on a given background).
 */

import type { Rgb, OnColorResult } from './types.js';
import { parseOklch } from './color-conv.js';

/**
 * Calculates WCAG 2.1 relative luminance from linear sRGB values.
 * @param r - Red channel (0-1, linear)
 * @param g - Green channel (0-1, linear)
 * @param b - Blue channel (0-1, linear)
 */
export function getLuminance(r: number, g: number, b: number): number {
    const a = [r, g, b].map((v) =>
        v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4)
    );
    return a[0]! * 0.2126 + a[1]! * 0.7152 + a[2]! * 0.0722;
}

/**
 * Calculates WCAG contrast ratio between two RGB colors.
 * @param rgb1 - First color (0-255 per channel)
 * @param rgb2 - Second color (0-255 per channel)
 * @returns Contrast ratio (1-21)
 */
export function getContrast(rgb1: Rgb, rgb2: Rgb): number {
    const lum1 = getLuminance(rgb1.r / 255, rgb1.g / 255, rgb1.b / 255);
    const lum2 = getLuminance(rgb2.r / 255, rgb2.g / 255, rgb2.b / 255);
    const brightest = Math.max(lum1, lum2);
    const darkest = Math.min(lum1, lum2);
    return (brightest + 0.05) / (darkest + 0.05);
}

/**
 * Automatically selects White or Black text for accessible contrast
 * on a given background color.
 *
 * - Returns White if contrast >= 4.5 and White >= Black
 * - Returns Black if contrast >= 4.5 and Black > White
 * - Falls back to the higher-contrast option with a warning
 *
 * @param baseOklchStr - OKLCH string of the background color
 */
export function deriveOnColor(baseOklchStr: string): OnColorResult {
    const parsed = parseOklch(baseOklchStr);
    if (!parsed) throw new Error(`Invalid OKLCH: ${baseOklchStr}`);

    const baseRgb: Rgb = { r: parsed.r, g: parsed.g, b: parsed.b };
    const whiteRgb: Rgb = { r: 255, g: 255, b: 255 };
    const blackRgb: Rgb = { r: 0, g: 0, b: 0 };

    const contrastWhite = getContrast(baseRgb, whiteRgb);
    const contrastBlack = getContrast(baseRgb, blackRgb);

    if (contrastWhite >= 4.5 && contrastWhite >= contrastBlack) {
        return { oklch: 'oklch(100% 0 0)', contrast: contrastWhite, color: 'White' };
    } else if (contrastBlack >= 4.5 && contrastBlack > contrastWhite) {
        return { oklch: 'oklch(0% 0 0)', contrast: contrastBlack, color: 'Black' };
    } else {
        // Fallback: pick the best one, even if < 4.5
        const best =
            contrastWhite > contrastBlack
                ? { oklch: 'oklch(100% 0 0)', contrast: contrastWhite, color: 'White' as const }
                : { oklch: 'oklch(0% 0 0)', contrast: contrastBlack, color: 'Black' as const };

        return { ...best, warning: true };
    }
}

File: dart/lib/src/contrast.dart
/// Contrast and automated accessibility rules.
import 'dart:math' as math;
import 'types.dart';
import 'color_conv.dart';

double getLuminance(double r, double g, double b) {
  double toLinear(double v) {
    return v <= 0.03928
        ? v / 12.92
        : math.pow((v + 0.055) / 1.055, 2.4).toDouble();
  }

  return 0.2126 * toLinear(r) + 0.7152 * toLinear(g) + 0.0722 * toLinear(b);
}

double getContrast(RgbaColor rgb1, RgbaColor rgb2) {
  final l1 = getLuminance(rgb1.r / 255.0, rgb1.g / 255.0, rgb1.b / 255.0);
  final l2 = getLuminance(rgb2.r / 255.0, rgb2.g / 255.0, rgb2.b / 255.0);
  final brighter = math.max(l1, l2);
  final darker = math.min(l1, l2);
  return (brighter + 0.05) / (darker + 0.05);
}

class OnColorResult {
  final String oklch;
  final double contrast;
  final String color;
  final bool warning;

  const OnColorResult({
    required this.oklch,
    required this.contrast,
    required this.color,
    this.warning = false,
  });
}

OnColorResult deriveOnColor(String baseOklchStr) {
  final parsed = parseOklch(baseOklchStr);
  if (parsed == null)
    throw ArgumentError('Invalid OKLCH string: \$baseOklchStr');

  final bgRgb = RgbaColor(r: parsed.r, g: parsed.g, b: parsed.b);
  final white = const RgbaColor(r: 255, g: 255, b: 255);
  final black = const RgbaColor(r: 0, g: 0, b: 0);

  final whiteContrast = getContrast(bgRgb, white);
  final blackContrast = getContrast(bgRgb, black);

  if (whiteContrast >= 4.5 && whiteContrast >= blackContrast) {
    return OnColorResult(
        oklch: 'oklch(100% 0 0)', contrast: whiteContrast, color: 'White');
  } else if (blackContrast >= 4.5 && blackContrast > whiteContrast) {
    return OnColorResult(
        oklch: 'oklch(0% 0 0)', contrast: blackContrast, color: 'Black');
  } else {
    if (whiteContrast >= blackContrast) {
      return OnColorResult(
          oklch: 'oklch(100% 0 0)',
          contrast: whiteContrast,
          color: 'White',
          warning: true);
    } else {
      return OnColorResult(
          oklch: 'oklch(0% 0 0)',
          contrast: blackContrast,
          color: 'Black',
          warning: true);
    }
  }
}
